week-1-ts-ecosystem/w01e05-my-own-cli/src/showcase-demo.ts

```diff
@@ -0,0 +1,184 @@
+import { program } from 'commander'
+import {
+  ActionTypes,
+  CartType,
+  chooseAction,
+} from './showcaseActions/chooseAction'
+import { chooseCart } from './showcaseActions/chooseCart'
+import {
+  AuctionProduct,
+  BuyNowProduct,
+  FreeProduct,
+} from './shared_w01e03/Products'
+import Table from 'cli-table'
+import { removeItems } from './showcaseActions/removeItems'
+import { addItems } from './showcaseActions/addItems'
+import { faker } from '@faker-js/faker/locale/pl'
+import { logger } from './utils'
+import { DataRepository } from './DataRepository'
+import { PersistantCart } from './PersistantCart'
+
+program
+  .name('demo-showcase')
+  .description('CLI for demo showcase')
+  .version('0.1.0')
+program.option('--persistence', 'saves data to json file')
+
+program.parse()
+
+const options: { persistence: boolean } = program.opts() || {
+  persistence: false,
+}
+
+logger.info(`Persistence mode is: ${options.persistence}`)
+
+const AuctionsDataRepository = new DataRepository<AuctionProduct>(
+  'auctions',
+  options.persistence
+)
+const BuyNowDataRepository = new DataRepository<BuyNowProduct>(
+  'buyNow',
+  options.persistence
+)
+const FreeCartRepository = new DataRepository<FreeProduct>(
+  'forFree',
+  options.persistence
+)
+
+const AuctionCart = new PersistantCart<AuctionProduct>(AuctionsDataRepository)
+const BuyNowCart = new PersistantCart<BuyNowProduct>(BuyNowDataRepository)
+const FreeCart = new PersistantCart<FreeProduct>(FreeCartRepository)
+
+const start = () =>
+  chooseCart(({ cartType }) => {
+    if (cartType === 'quit') {
+      return
+    }
+
+    return chooseAction(cartType, handleAction)
+  })
+
+const displayCart = (cart: CartType) => {
+  let table: Table
+  switch (cart) {
+    case 'forFree':
```

Comment:
Generalnie całe wykonanie to dobra robota.
Ładnie zorganizowane w oddzielne pytania do `EndUser`'a "actions"

Nie ma misz-maszu. Struktura ogarnięta dobrze.

Jedyne miejsce które docelowo bym zrefaktoryzował to właśnie te elementy gdzie robisz `switch` po `CartType`

Wydaje mi się że da się rozbudować `PersistantCart` tak aby był dedykowany (być może wyjdą Ci dodatkowe 3 klasy) ale wtedy to tam zajmowałbyś się tematami ogarniania tego jak i co i gdzie mapować etc. i jak to wciągać.

Chodzi o to że w momencie gdy user wybiera że chodzi o np. `BuyNowProduct`.

Ty już masz wybrane: `BuyNowCart` i to na nim operujesz. nie musisz robić `switch` bo każdy `Cart` miałby ten sam interfejs.

Wtedy dużo logiki stąd, weszłoby do `Cart`'ów a pozbył byś sie `switch`y